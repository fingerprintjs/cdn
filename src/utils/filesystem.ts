import * as path from 'path'
import * as os from 'os'
import { createReadStream, promises as fs } from 'fs'

export function makeTemporaryDirectory(): Promise<string> {
  return fs.mkdtemp(path.join(os.tmpdir(), 'fpjs-'))
}

export async function readFirstCharacters(filePath: string, maxLength: number): Promise<string> {
  let fileContent = ''
  const readStream = createReadStream(filePath, { encoding: 'utf8', highWaterMark: maxLength })

  try {
    await new Promise<void>((resolve, reject) => {
      readStream.on('data', (chunk) => {
        fileContent += chunk
        if (fileContent.length >= maxLength) {
          fileContent = fileContent.slice(0, maxLength)
          resolve()
        }
      })
      readStream.on('error', reject)
      readStream.on('end', resolve)
    })
  } finally {
    readStream.close()
  }

  return fileContent
}

export async function withTemporaryFiles<T>(
  files: Record<string, string>,
  action: (directory: string) => Promise<T> | T,
): Promise<T> {
  const directory = await makeTemporaryDirectory()
  try {
    await Promise.all(
      Object.entries(files).map(async ([filePath, content]) => {
        const fileFullPath = path.join(directory, filePath)
        await fs.mkdir(path.dirname(fileFullPath), { recursive: true })
        await fs.writeFile(fileFullPath, content)
      }),
    )
    return await action(directory)
  } finally {
    await fs.rm(directory, { recursive: true, force: true })
  }
}

export async function withTemporaryFile<T>(content: string, action: (file: string) => Promise<T> | T): Promise<T> {
  const file = path.join(os.tmpdir(), `fpjs-${Math.random().toString(36).slice(2)}.txt`)

  try {
    await fs.writeFile(file, content)
    return await action(file)
  } finally {
    await fs.rm(file, { force: true })
  }
}

export async function isDirectory(path: string): Promise<boolean> {
  try {
    const fileInfo = await fs.lstat(path)
    return fileInfo.isDirectory()
  } catch (error) {
    // Means that the file doesn't exist
    if ((error as NodeJS.ErrnoException).code === 'ENOENT') {
      return false
    }
    throw error
  }
}
