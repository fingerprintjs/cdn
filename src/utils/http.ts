import { CloudFrontHeaders, CloudFrontResultResponse } from 'aws-lambda'

type Status = Pick<CloudFrontResultResponse, 'status' | 'statusDescription'>

export const okStatus: Status = {
  status: '200',
}

export const permanentRedirectStatus: Status = {
  status: '301',
  statusDescription: 'Moved Permanently',
}

export const temporaryRedirectStatus: Status = {
  status: '302',
  statusDescription: 'Found',
}

export const notFoundStatus: Status = {
  status: '404',
  statusDescription: 'Not Found',
}

export function makeCorsHeaders(): CloudFrontHeaders {
  return {
    'access-control-allow-origin': [{ value: '*' }],
  }
}

/**
 * Notice: the durations are set in milliseconds like everything else in JavaScript
 */
export function makeCacheHeaders(
  browserCacheDuration: number,
  cdnCacheDuration = browserCacheDuration,
  durationFluctuation = 0.1,
): CloudFrontHeaders {
  // todo: Consider adding ETag
  const headers: CloudFrontHeaders = {}
  const fluctuationMultiplier = 1 - durationFluctuation / 2 + durationFluctuation * Math.random()
  const cacheControl = ['public', `max-age=${((browserCacheDuration / 1000) * fluctuationMultiplier).toFixed()}`]

  if (browserCacheDuration !== cdnCacheDuration) {
    cacheControl.push(`s-maxage=${applyFluctuation(cdnCacheDuration / 1000, durationFluctuation).toFixed()}`)

    // When the browser cache life is longer than the CDN cache life, the "If-Modified" check will always return "true",
    // so the Last-Modified header affects nothing.
    if (browserCacheDuration < cdnCacheDuration) {
      headers['last-modified'] = [{ value: new Date().toUTCString() }]
    }
  }

  headers['cache-control'] = [{ value: cacheControl.join(', ') }]
  return headers
}

function applyFluctuation(value: number, fluctuation: number): number {
  return value * (1 - fluctuation / 2 + fluctuation * Math.random())
}
