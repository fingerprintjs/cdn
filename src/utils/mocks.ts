import { CloudFrontRequestEvent, Context } from 'aws-lambda'
import * as zlib from 'zlib'
import * as tarStream from 'tar-stream'
import { RegistryPackageShortData } from '../npm'

export function makeMockCloudFrontEvent(uri: string): CloudFrontRequestEvent {
  return {
    Records: [
      {
        cf: {
          config: {
            distributionDomainName: 'd111111abcdef8.cloudfront.net',
            distributionId: 'EDFDVBD6EXAMPLE',
            eventType: 'origin-request',
            requestId: '4TyzHTaYWb1GX1qTfsHhEqV6HUDd_BzoBZnwfnvQc_1oF26ClkoUSEQ==',
          },
          request: {
            clientIp: '203.0.113.178',
            headers: {},
            method: 'GET',
            querystring: '',
            uri,
          },
        },
      },
    ],
  }
}

export function makeMockLambdaContext() {
  return {} as unknown as Context
}

/**
 * A mock response data about a package from the NPM registry
 */
export const mockNpmPackageInfo: RegistryPackageShortData = {
  name: '@fpjs-incubator/botd-agent',
  'dist-tags': {
    latest: '0.3.0',
    dev: '0.2.1-dev.0',
  },
  versions: {
    '0.1.15': {
      name: '@fpjs-incubator/botd-agent',
      version: '0.1.15',
      dependencies: { tslib: '^2.0.1' },
      devDependencies: { '@rollup/plugin-json': '^4.1.0' },
      dist: {
        shasum: 'd10ae3f29b70236703dc266094f999134ec9c846',
        integrity: 'sha512-voyRD+FrqLWJksQyB94OELKYNkL67MC5Wx7g9VHfhPDIC+SMgI84L2c4rynO/eSHqV2LVfPKqpR+T0oRE/+7ow==',
        tarball: 'https://registry.npmjs.org/@fpjs-incubator/botd-agent/-/botd-agent-0.1.15.tgz',
        fileCount: 12,
        unpackedSize: 476446,
        'npm-signature': '...',
      },
    },
    '0.1.16': {
      name: '@fpjs-incubator/botd-agent',
      version: '0.1.16',
      dependencies: { tslib: '^2.0.1' },
      devDependencies: { '@rollup/plugin-json': '^4.1.0' },
      dist: {
        shasum: 'dc327528121dee3346adbe59c6776d0a42cd0a6f',
        integrity: 'sha512-W+lqReWjLFB0rvpX96G/0QoM+vD4563aAWAsUTWTrQ8eL8NSsfxmPyqhis8bZsp/enlLtH/Iii0J2e5RNl2GxA==',
        tarball: 'https://registry.npmjs.org/@fpjs-incubator/botd-agent/-/botd-agent-0.1.16.tgz',
        fileCount: 12,
        unpackedSize: 480467,
        'npm-signature': '...',
      },
    },
    '0.2.0': {
      name: '@fpjs-incubator/botd-agent',
      version: '0.2.0',
      dependencies: { tslib: '^2.0.1' },
      devDependencies: { '@rollup/plugin-json': '^4.1.0' },
      dist: {
        shasum: 'de2f5abef524331db22b1745858e1b22bf5e86c6',
        integrity: 'sha512-uvjJIRLrDPHV9jA5+1X23o6o+PpxVzre/uZcrfuTBRnQMWM7lWkgj92GWeCG2GZKkkIbKTdAemWc0u5V7A5U8Q==',
        tarball: 'https://registry.npmjs.org/@fpjs-incubator/botd-agent/-/botd-agent-0.2.0.tgz',
        fileCount: 12,
        unpackedSize: 505276,
        'npm-signature': '...',
      },
    },
    '0.2.1': {
      name: '@fpjs-incubator/botd-agent',
      version: '0.2.1',
      dependencies: { tslib: '^2.0.1' },
      devDependencies: { '@rollup/plugin-json': '^4.1.0' },
      dist: {
        shasum: '936c7d7fd563dd72428471ea7b98cb6607b7f3af',
        integrity: 'sha512-zXkfhFHnNYMXclLSYldKnsANRFDiHlcA2ZmRnDSp3D3KSE8pNVh/Ttwscc3JoNSBjSds9DO2TJYLNhAI+vJJJQ==',
        tarball: 'https://registry.npmjs.org/@fpjs-incubator/botd-agent/-/botd-agent-0.2.1.tgz',
        fileCount: 12,
        unpackedSize: 505680,
        'npm-signature': '...',
      },
    },
    '0.2.1-dev.0': {
      name: '@fpjs-incubator/botd-agent',
      version: '0.2.1-dev.0',
      dependencies: { tslib: '^2.0.1' },
      devDependencies: { '@rollup/plugin-json': '^4.1.0' },
      dist: {
        shasum: '936c7d7fd563dd72428471ea7b98cb6607b7f3af',
        integrity: 'sha512-zXkfhFHnNYMXclLSYldKnsANRFDiHlcA2ZmRnDSp3D3KSE8pNVh/Ttwscc3JoNSBjSds9DO2TJYLNhAI+vJJJQ==',
        tarball: 'https://registry.npmjs.org/@fpjs-incubator/botd-agent/-/botd-agent-0.2.1-dev.0.tgz',
        fileCount: 12,
        unpackedSize: 505680,
        'npm-signature': '...',
      },
    },
    '0.3.0': {
      name: '@fpjs-incubator/botd-agent',
      version: '0.3.0',
      dependencies: { tslib: '^2.0.1' },
      devDependencies: { '@rollup/plugin-json': '^4.1.0' },
      dist: {
        shasum: 'd15c2fab8e0d9138bcb8efa1429dd780f93bb945',
        integrity: 'sha512-2kw2yVrfMKd0YvmYAdNxhGytNhhLvChqNAbBZWMglYVw2J95Jm50ketK5yS1C8SJmqG2nFJgeuWvYXeZEh+b8g==',
        tarball: 'https://registry.npmjs.org/@fpjs-incubator/botd-agent/-/botd-agent-0.3.0.tgz',
        fileCount: 12,
        unpackedSize: 524092,
        'npm-signature': '...',
      },
    },
  },
  modified: '2021-07-29T03:10:07.866Z',
}

/**
 * Makes a readable stream with an NPM package archive. You can use it with `nock` wo mock an NPM response
 */
export function mockNpmPackageDownloadStream(files: Record<string, string>): zlib.Gzip {
  const pack = tarStream.pack()
  for (const [filePath, content] of Object.entries(files)) {
    pack.entry({ name: `package/${filePath}` }, content)
  }
  pack.finalize()
  return pack.pipe(zlib.createGzip())
}
