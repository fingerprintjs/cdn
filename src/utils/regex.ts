export function escapeForRegex(str: string): string {
  return str.replace(/[-[\]/{}()*+?.\\^$|]/g, '\\$&')
}
