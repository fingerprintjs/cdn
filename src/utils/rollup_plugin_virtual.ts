/*
 * This code is copied from https://github.com/rollup/plugins/blob/master/packages/virtual/src/index.ts
 *
 * When this issue is solved: https://github.com/rollup/plugins/issues/894,
 * this file should be replaced with the NPM package: @rollup/plugin-virtual
 */

import * as path from 'path'
import { Plugin } from 'rollup'

export interface RollupVirtualOptions {
  [id: string]: string
}

const PREFIX = `\0virtual:`

export default function virtual(modules: RollupVirtualOptions): Plugin {
  const resolvedIds = new Map<string, string>()

  Object.keys(modules).forEach((id) => {
    resolvedIds.set(path.resolve(id), modules[id])
  })

  return {
    name: 'virtual',

    resolveId(id, importer) {
      if (id in modules) return PREFIX + id

      if (importer) {
        const importerNoPrefix = importer.startsWith(PREFIX) ? importer.slice(PREFIX.length) : importer
        const resolved = path.resolve(path.dirname(importerNoPrefix), id)
        if (resolvedIds.has(resolved)) return PREFIX + resolved
      }

      return null
    },

    load(id) {
      if (id.startsWith(PREFIX)) {
        const idNoPrefix = id.slice(PREFIX.length)

        return idNoPrefix in modules ? modules[idNoPrefix] : resolvedIds.get(idNoPrefix)
      }

      return null
    },
  }
}
