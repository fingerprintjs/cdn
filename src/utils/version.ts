/**
 * Compares version strings.
 * Returns:
 * -1 - the first version is lower than the second version
 *  0 - the versions are equal
 *  1 - the first version is greater than the second version
 */
export function compareVersions(version1: string, version2: string): -1 | 0 | 1 {
  const [firstPart1, rest1] = extractFirstVersionPart(version1)
  const [firstPart2, rest2] = extractFirstVersionPart(version2)

  const firstNumber1 = parseInt(firstPart1)
  const firstNumber2 = parseInt(firstPart2)
  if (isNaN(firstNumber1) && isNaN(firstNumber2)) {
    return 0
  }
  if (isNaN(firstNumber1)) {
    return -1
  }
  if (isNaN(firstNumber2)) {
    return 1
  }
  if (firstNumber1 !== firstNumber2) {
    return firstNumber1 < firstNumber2 ? -1 : 1
  }

  // The first numbers of the versions are equal, so comparing the rest parts of the versions
  if (rest1 === undefined && rest2 === undefined) {
    return 0
  }
  if (rest1 === undefined) {
    return -1
  }
  if (rest2 === undefined) {
    return 1
  }
  return compareVersions(rest1, rest2)
}

/**
 * Checks whether the version lies in the version range.
 * The start version is included and the end version is NOT included.
 */
export function isVersionInRange(
  startVersion: string | undefined,
  versionToCheck: string,
  endVersion: string | undefined,
): boolean {
  return (
    (startVersion === undefined || compareVersions(startVersion, versionToCheck) <= 0) &&
    (endVersion === undefined || compareVersions(versionToCheck, endVersion) < 0)
  )
}

/**
 * Checks whether the actualVersion can be used as the expectedVersion according to the Semantic Versioning rules
 */
export function doesVersionMatch(expectedVersion: string, actualVersion: string): boolean {
  if (expectedVersion === actualVersion) {
    return true
  }
  if (!/^[\d.]+$/.test(actualVersion)) {
    // Special versions with `-dev`, `-beta`, etc never match the expected version only in case of full match
    return false
  }
  const [expectedMajorVersion, expectedVersionRest = '0'] = extractFirstVersionPart(expectedVersion)
  const [actionMajorVersion, actionVersionRest = '0'] = extractFirstVersionPart(actualVersion)
  if (expectedMajorVersion !== actionMajorVersion) {
    return false
  }
  if (parseInt(expectedMajorVersion) === 0) {
    // When the major version is 0, the minor version is treated like a major version too
    return doesVersionMatch(expectedVersionRest, actionVersionRest)
  }
  return compareVersions(expectedVersionRest, actionVersionRest) <= 0
}

function extractFirstVersionPart(version: string): [first: string, rest?: string] {
  const pointIndex = version.indexOf('.')
  return pointIndex === -1 ? [version] : [version.slice(0, pointIndex), version.slice(pointIndex + 1)]
}
