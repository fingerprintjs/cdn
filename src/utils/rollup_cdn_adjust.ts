import type { Plugin } from 'rollup'
import { escapeForRegex } from './regex'

interface Options {
  /**
   * The words to replace in the library distributive files. The keys are the targets, the values are the replacements
   * (inserted as is). When searching, the targets are expected to be surrounded by word boundaries (\b in regexps).
   */
  replacements?: Record<string, string> | undefined
}

/**
 * Creates a Rollup plugin for CDN-specific transformations
 */
export default function rollupCdnAdjust({ replacements }: Options = {}): Plugin {
  // Official guide: https://rollupjs.org/guide/en/#transformers
  // Inspired by: https://github.com/rollup/plugins/blob/master/packages/replace/src/index.js
  return {
    name: 'cdnAdjust',

    transform(source, id) {
      if (!replacements || Object.keys(replacements).length === 0 || isThirdPartyLibrary(id)) {
        return null
      }
      return replaceInCode(source, replacements)
    },
  }
}

function isThirdPartyLibrary(id: string) {
  return id.includes('/node_modules/')
}

function replaceInCode(code: string, replacements: Record<string, string>) {
  return code.replace(
    new RegExp(`(?<=\\W)(${Object.keys(replacements).map(escapeForRegex).join('|')})(?=\\W)`, 'g'),
    (key) => replacements[key],
  )
}
