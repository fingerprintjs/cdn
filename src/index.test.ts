import * as nock from 'nock'
import { handler } from './index'
import * as mocks from './utils/mocks'

/*
 * These tests make no real HTTP requests. The tests with real requests are the "integration" tests.
 */

beforeAll(() => {
  nock.disableNetConnect()
})

afterAll(() => {
  nock.enableNetConnect()
})

afterEach(() => {
  nock.cleanAll()
})

function callHandler(uri: string) {
  return handler(mocks.makeMockCloudFrontEvent(uri), mocks.makeMockLambdaContext(), () => undefined)
}

it('makes the root page', async () => {
  const response = await callHandler('/')
  expect(response).toEqual({
    status: '200',
    headers: {
      'cache-control': [{ value: expect.stringMatching(/^\s*public,\s*max-age=\d+\s*$/) }],
      'access-control-allow-origin': [{ value: '*' }],
      'strict-transport-security': [{ value: 'max-age=63072000; includeSubDomains; preload' }],
      'content-type': [{ value: 'text/plain; charset=utf-8' }],
      'x-content-type-options': [{ value: 'nosniff' }],
    },
    body: 'This is a FingerprintJS CDN',
  })
})

it('returns "not found" error', async () => {
  const response = await callHandler('/not/existing/page')
  expect(response).toEqual({
    status: '404',
    statusDescription: 'Not Found',
    headers: {
      'cache-control': [{ value: expect.stringMatching(/^\s*public,\s*max-age=\d+,\s*s-maxage=\d+\s*$/) }],
      'access-control-allow-origin': [{ value: '*' }],
      'strict-transport-security': [{ value: 'max-age=63072000; includeSubDomains; preload' }],
      'content-type': [{ value: 'text/plain; charset=utf-8' }],
      'x-content-type-options': [{ value: 'nosniff' }],
    },
    body: "The /not/existing/page path doesn't exist",
  })
})

describe('inexact version', () => {
  beforeEach(() => {
    nock('https://registry.npmjs.org', {
      reqheaders: {
        Accept: 'application/vnd.npm.install-v1+json',
      },
    })
      .get('/@fingerprintjs/fingerprintjs')
      .reply(200, mocks.mockNpmPackageInfo)
  })

  it('handles missing version', async () => {
    const response = await callHandler('/fingerprintjs/v3.0')
    expect(response).toEqual({
      status: '404',
      statusDescription: 'Not Found',
      headers: expect.objectContaining({}),
      body: 'There is no version matching 3.0.*',
    })
  })

  it('redirects to the exact version', async () => {
    const response = await callHandler('/fingerprintjs/v3.2/umd.js')
    expect(response).toEqual({
      status: '302',
      statusDescription: 'Found',
      headers: expect.objectContaining({
        location: [{ value: '/fingerprintjs/v3.2.1/umd.js' }],
      }),
    })
  })

  it('follows the route redirect', async () => {
    const response = await callHandler('/fingerprintjs/v3.2')
    expect(response).toEqual({
      status: '302',
      statusDescription: 'Found',
      headers: expect.objectContaining({
        location: [{ value: '/fingerprintjs/v3.2.1/esm.min.js' }],
      }),
    })
  })
})

describe('exact version', () => {
  it('redirects', async () => {
    const response = await callHandler('/fingerprintjs/v3.0.1')
    expect(response).toEqual({
      status: '301',
      statusDescription: 'Moved Permanently',
      headers: expect.objectContaining({
        location: [{ value: '/fingerprintjs/v3.0.1/esm.min.js' }],
      }),
    })
  })

  it('monitors', async () => {
    const response = await callHandler('/fingerprintjs/v3.0.1/npm-monitoring')
    expect(response).toEqual({
      status: '200',
      headers: expect.objectContaining({}),
    })
  })

  it('handles missing version', async () => {
    nock('https://registry.npmjs.org').get('/@fingerprintjs/fingerprintjs/-/fingerprintjs-3.2.1.tgz').reply(404)
    const response = await callHandler('/fingerprintjs/v3.2.1/esm.js')
    expect(response).toEqual({
      status: '404',
      statusDescription: 'Not Found',
      headers: expect.objectContaining({}),
      body: 'There is no version 3.2.1',
    })
  })

  const packageFiles = {
    'package.json': '{"module": "index.js"}',
    'index.js': `import {__assign} from 'tslib'
export function test() {
  if (!window.__fpjs_d_m) {
    console.log('Test') // Shall be removed
  }
  return __assign({}, {})
}`,
  }

  it('builds unminified UMD with replacements', async () => {
    nock('https://registry.npmjs.org')
      .get('/@fingerprintjs/fingerprintjs/-/fingerprintjs-3.2.1.tgz')
      .reply(200, mocks.mockNpmPackageDownloadStream(packageFiles), {
        'Content-Type': 'application/octet-stream',
      })
    const response = await callHandler('/fingerprintjs/v3.2.1/umd.js')
    expect(response).toEqual({
      status: '200',
      headers: expect.objectContaining({
        'content-type': [{ value: 'text/javascript; charset=utf-8' }],
      }),
      body: expect.anything(),
    })
    expect(response?.body).toMatchSnapshot()
  })

  it('builds minified ESM', async () => {
    nock('https://registry.npmjs.org')
      .get('/@fpjs-incubator/botd-agent/-/botd-agent-0.1.20.tgz')
      .reply(200, mocks.mockNpmPackageDownloadStream(packageFiles), {
        'Content-Type': 'application/octet-stream',
      })
    const response = await callHandler('/botd/v0.1.20/esm.min.js')
    expect(response).toEqual({
      status: '200',
      headers: expect.objectContaining({}),
      body: expect.anything(),
    })
    expect(response?.body).toMatchSnapshot()
  })
})
