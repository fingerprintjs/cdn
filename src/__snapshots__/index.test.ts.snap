// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`exact version builds minified ESM 1`] = `
"var n=function(){return n=Object.assign||function(n){for(var r,t=1,o=arguments.length;t<o;t++)for(var e in r=arguments[t])Object.prototype.hasOwnProperty.call(r,e)&&(n[e]=r[e]);return n},n.apply(this,arguments)};function r(){return window.__fpjs_d_m||console.log("Test"),n({},{})}export{r as test};
"
`;

exports[`exact version builds minified ESM 2`] = `
[
  {
    "value": ""dBhlDEIPIeluHpJTNf6LmEsi+1c"",
  },
]
`;

exports[`exact version builds unminified UMD with replacements 1`] = `
"(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports) :
    typeof define === 'function' && define.amd ? define(['exports'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global.FingerprintJS = {}));
})(this, (function (exports) { 'use strict';

    /******************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    function test() {
      return __assign({}, {})
    }

    exports.test = test;

}));
"
`;

exports[`exact version builds unminified UMD with replacements 2`] = `
[
  {
    "value": ""HyvbPZ1n88H9T4PTPQ2W/FrdS60"",
  },
]
`;
